sky/provision/aws/instance.py

```diff
@@ -0,0 +1,132 @@
+            if inst.id not in excluded_instances:
+                included_instances.append(inst.id)
+        instances = instances.filter(InstanceIds=included_instances)
+    return instances
+
+
+def stop_instances(region: str,
+                   cluster_name: str,
+                   included_instances: Optional[List[str]] = None,
+                   excluded_instances: Optional[List[str]] = None) -> None:
+    """See sky/provision/__init__.py"""
+    ec2 = utils.create_resource('ec2', region=region)
+    filters = [
+        {
+            'Name': 'instance-state-name',
+            'Values': ['pending', 'running'],
+        },
+        {
+            'Name': f'tag:{TAG_RAY_CLUSTER_NAME}',
+            'Values': [cluster_name],
+        },
+    ]
+    # TODO(suquark): wait pending clusters?
+    instances = _filter_instances(ec2, filters, included_instances,
+                                  excluded_instances)
+    instances.stop()
+
+
+def terminate_instances(region: str,
+                        cluster_name: str,
+                        included_instances: Optional[List[str]] = None,
+                        excluded_instances: Optional[List[str]] = None) -> None:
+    """See sky/provision/__init__.py"""
+    ec2 = utils.create_resource('ec2', region=region)
+    filters = [
+        {
+            'Name': 'instance-state-name',
+            # exclude 'shutting-down' or 'terminated' states
+            'Values': ['pending', 'running', 'stopping', 'stopped'],
+        },
+        {
+            'Name': f'tag:{TAG_RAY_CLUSTER_NAME}',
+            'Values': [cluster_name],
+        },
+    ]
+    # TODO(suquark): wait pending clusters?
+    # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Instance
+    instances = _filter_instances(ec2, filters, included_instances,
+                                  excluded_instances)
+    instances.terminate()
+
+
+def wait_instances(region: str, cluster_name: str,
+                   state: common.InstanceState) -> None:
+    """See sky/provision/__init__.py"""
+    # possible exceptions: https://github.com/boto/boto3/issues/176
+    ec2 = utils.create_resource('ec2', region=region)
+    client = ec2.meta.client
+
+    filters = [
+        {
+            'Name': f'tag:{TAG_RAY_CLUSTER_NAME}',
+            'Values': [cluster_name],
+        },
+    ]
+
+    if state == common.InstanceState.RUNNING:
+        # NOTE: there could be a terminated/terminating AWS cluster with
+        # the same cluster name.
+        # Waiting the cluster results in errors (cannot wait for 'terminated').
+        # So here we exclude terminated/terminating instances.
+        filters.append({
+            'Name': 'instance-state-name',
+            'Values': ['pending', 'running'],
+        })
+    elif state == common.InstanceState.STOPPED:
+        filters.append({
+            'Name': 'instance-state-name',
+            'Values': ['stopping', 'stopped'],
```

Comment:
Question: why we only wait for instance with `stopping` and `stopped` states for `state == common.InstanceState.STOPPED`, but waiting for all the instances when `state == common.InstanceState.TERMINATED`? Should we wait for all the instance except for the `terminating` and `terminated` ones to be stopped? Similar for the `RUNNING`, should we wait for the `stopped` `stopping` instances as well?

Reply:
Got it! That makes sense. Thanks for the explanation